dali/pipeline/data/buffer_manager.h

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_DATA_BUFFER_MANAGER_H_
+#define DALI_PIPELINE_DATA_BUFFER_MANAGER_H_
+
+#include <map>
+#include <memory>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "dali/pipeline/operators/op_spec.h"
+#include "dali/pipeline/data/allocator_manager.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/buffer.h"
+
+namespace dali {
+
+#define LINEAR_SEARCH
+
+// Linear search over available buffers to get the smallest available that
+// is >= bytes requested
+template <typename Backend>
+inline unique_ptr<Buffer<Backend>> LinearSearch(vector<unique_ptr<Buffer<Backend>>> *buffers,
+                                                size_t nbytes,
+                                                bool pinned = false) {
+  int current_idx = -1;
+  size_t current_delta = (size_t) -1;
+
+  for (size_t i = 0; i < buffers->size(); ++i) {
+    const size_t buffer_size = (*buffers)[i]->capacity();
+    if (buffer_size >= 0.5 * nbytes &&
```

Comment:
Why we consider something that is just 0.5 * requested_size as a fit?

Reply:
We need to make some heuristic here and 0.5x-2x is just first stab at it. Basically if you need 1GB and have 0.9GB buffer already, you don't want to allocate a completely new 1GB buffer and instead take the smaller one and resize.